universal/include/userver/formats/universal/universal.hpp

```diff
@@ -0,0 +1,357 @@
+constexpr inline auto RunParseCheckFor(const Value&, std::optional<Field>& field, Default<DefaultValue>) {
+  if(!field.has_value()) {
+    field = DefaultValue;
+  };
+};
+
+
+template <typename T, auto I, typename... Params, typename Builder, typename Value>
+constexpr inline std::enable_if_t<utils::impl::AnyOf(utils::impl::IsSameCarried<Additional>(), utils::impl::TypeList<Params...>{}), void>
+RunWrite(Builder& builder, const std::unordered_map<std::string, Value>& field) {
+  for(const auto& element : field) {
+    builder[element.first] = element.second;
+  };
+};
+
+template <typename T, auto I, typename Builder, typename Field>
+constexpr inline auto RunCheckFor(Builder&, Field&&, Disabled) noexcept {};
+
+template <typename T, auto I, typename Builder, typename Field>
+constexpr inline auto RunCheckFor(Builder&, const std::optional<Field>&, Disabled) noexcept {};
+
+template <typename T, auto I, typename Builder, typename Field, typename CheckT>
+constexpr inline auto RunCheckFor(Builder&, Field&& field, CheckT check) {
+  if(!Check(field, check)) {
+    throw std::runtime_error(ErrorMessage<T, I>(std::forward<Field>(field), check));
+  };
+};
+
+} // namespace exam
+template <typename T, auto I, typename Builder, typename... Params>
+constexpr inline auto UniversalSerializeField(
+     FieldParametries<T, I, Params...>
+    ,Builder& builder
+    ,const T& obj) {
+  const auto& value = boost::pfr::get<I>(obj);
+  using exam::RunCheckFor;
+  using exam::RunWrite;
+  (RunCheckFor<T, I>(builder, value, Params{}), ...);
+  RunWrite<T, I, Params...>(builder, value);
+};
+
+template <typename T, auto I, typename Format, typename... Params>
+constexpr inline auto UniversalParseField(
+     FieldParametries<T, I, Params...>
+    ,Format&& from) {
+  using exam::Read;
+  using exam::RunParseCheckFor;
+  using FieldType = std::remove_cvref_t<decltype(boost::pfr::get<I>(std::declval<T>()))>;
+  auto value = Read<T, I, Params...>(from, userver::formats::parse::To<FieldType>{});
+  (RunParseCheckFor<T, I>(from, value, Params{}), ...);
+  return value;
+};
+
+template <typename T>
+consteval bool AdditionalCheck(const T&) {
+  return false;
+};
+
+template <typename T>
+consteval bool AdditionalCheck(const std::optional<std::optional<T>>&) {
+  return true;
+};
+
+template <typename T, auto I, typename Format, typename... Params>
+constexpr inline std::optional<std::remove_cvref_t<decltype(boost::pfr::get<I>(std::declval<T>()))>>
+UniversalTryParseField(
+     FieldParametries<T, I, Params...>
+    ,Format&& from) noexcept {
+  using FieldType = std::remove_cvref_t<decltype(boost::pfr::get<I>(std::declval<T>()))>;
+  using exam::Check;
+  using exam::Read;
+
+
+  auto val = Read<T, I, Params...>(from, userver::formats::parse::To<std::optional<FieldType>>{});
+  if((Check(val, Params{}) && ...)) {
+    return val;
+  };
+  return std::nullopt;
+};
+
+} // namespace impl
+
+template <typename T, typename... Params>
+class SerializationConfig {
```

Comment:
[big] After some internal discussion: we want to reduce the amount of template metaprogramming and make `SerializationConfig` value-based. Details of struct fields serialization and so on will be stored inside `SerializationConfig` itself, not inside its template parameters.

`SerializationConfig` should contain a tuple of `FieldConfig` for each field. `FieldConfig` should be parametrized with the parsed type. The config creation will look like this:

```
.With<"field">({.max_items = 10, .items = {.min = 42}})
```

`With` takes a single parameter of type `FieldConfig<(type of "field")>`, which is a `struct` with all `optional` members, like `std::optional<std::size_t> max_items` and `FieldConfig<(type of items)> items`. This config is assigned to the `SerializationConfig`'s tuple element corresponding to `"field"`.

`FieldConfig` should be able to be specialized for the parsed type. For example, `FieldConfig` will be specialized for array-like ranges to include `min_items`, `max_items` and `items` fields, and for numeric types to include `min` and `max` fields of appropriate type.

`FieldConfig` for structs should contain a single **reference** to the struct's `SerializationConfig`, which should be default-initialized with `formats::universal::kSerialization<T>`. The "reference" part is important, as it allows to create recursive structs.

Add a test that contains a simple recursive type:

```cpp
struct Node {
  int value;
  std::vector<Node> children;
};
```

Test universal serialization on it.

That's the gist of the vision for value-based serialization config. Feel free to ask questions here or in PM.

Reply:
Functions will still be able to use `if constexpr` and alike by passing around `const FieldConfig<T>&` in template parameters. Remember that all those `SerializationConfig` and `FieldConfig` are stored inside variable templates, so it's possible to use addresses of those variables and their contents in templates, even in C++17